### PR TITLE
Ongoing fixes for 4.2.0 release

### PR DESCRIPTION
**Pull request by dhensby**
I'm putting my ongoing fixes to get 4.2.0 to release here.

### CODE REVIEW

File: tests/Model/Release/ComposerConstraintTest.php [hunk 1]
```diff
@@ -120,6 +120,10 @@ public function testParseDev()
         $constraint = new ComposerConstraint('^4.1@dev');
         $this->assertEquals('4.1.0-alpha1', $constraint->getMinVersion()->getValue());
         $this->assertEquals('4.99999.99999', $constraint->getMaxVersion()->getValue());
+
+        $constraint = new ComposerConstraint('4.2.0.x-dev');
+        $this->assertEquals('4.2.0-alpha1', $constraint->getMinVersion()->getValue());
+        $this->assertEquals('4.2.99999', $constraint->getMaxVersion()->getValue());
```
Comment (dhensby) [inline tests/Model/Release/ComposerConstraintTest.php]: this feels wrong to me and the max version should be 4.2.0, however the `ComposerConstraint::parse` method is unable to make any distinction about an explicit `.0` patch version and an implicit one, meaning we have to do 4.2.99999.

I'll run some local tests with cow to see how this works

Comment (robbieaverill) [reply tests/Model/Release/ComposerConstraintTest.php]: If it helps, composer does something like this too

Comment (tractorcow) [reply tests/Model/Release/ComposerConstraintTest.php]: max version should be 4.2.0.999999999. That's how composer internally translates `.x-dev`. It's missing the '0'

File: src/Model/Release/ComposerConstraint.php [hunk 2]
```diff
@@ -185,13 +185,23 @@ public static function parse($version)
...
         if ($valid) {
-            $minor = (isset($matches['minor']) && strlen($matches['minor'])) ? $matches['minor'] : '0';
-            $patch = (isset($matches['minor']) && $matches['minor']) ? '0' : null;
+            $minor = '0';
+            $patch = null;
+            if (isset($matches['minor'])) {
+                $minor = strlen($matches['minor']) ? $matches['minor'] : '0';
+                if ($matches['minor']) {
+                    if (isset($matches['patch']) && strlen($matches['patch'])) {
+                        $patch = $matches['patch'];
```
Comment (tractorcow) [inline src/Model/Release/ComposerConstraint.php]: sure only you'd need
```php
$patch = isset($matches['patch'])
    ? $matches['patch']
    : (isset($matches['minor']) && $matches['minor']) ? '0' : null; // existing logic
```

File: tests/Model/Release/ComposerConstraintTest.php (hunk 1 above)
Comment (dhensby) [reply tests/Model/Release/ComposerConstraintTest.php]: Yeah, ok - I think @tractorcow is right - there's just no way with our current API to discern if the patch number was explicit or implied :(

Comment (tractorcow) [reply tests/Model/Release/ComposerConstraintTest.php]: Maybe look at branching strategy?

**Comment by robbieaverill** [comment]
Do we think this pull request is worth reviving?

**Comment by robbieaverill** [comment]
Going to close this for now

**Pull request closed without merging**